src/app/shared/title-editor/title-editor.component.ts

```diff
@@ -0,0 +1,139 @@
+import { Component, ElementRef, EventEmitter, Input, OnInit, Output, ViewChild } from '@angular/core';
+import { AbstractControl, FormGroup, Validators } from '@angular/forms';
+import { UndoRedo } from '../../core/models/undoredo.model';
+
+const ISSUE_BODY_SIZE_LIMIT = 256;
+
+type textEntry = {
+  text: string;
+  selectStart: number;
+  selectEnd: number;
+};
+
+@Component({
+  selector: 'app-title-editor',
+  templateUrl: './title-editor.component.html',
+  styleUrls: ['./title-editor.component.css']
+})
+export class TitleEditorComponent implements OnInit {
+
+  constructor() {}
+
+  @Input() titleField: AbstractControl; // Compulsory Input
+  @Input() titleForm: FormGroup; // Compulsory Input
+  @Input() id: string; // Compulsory Input
+
+  @Input() initialTitle?: string;
+  placeholderText = 'Title';
+
+  @ViewChild('titleInput', { static: true }) titleInput: ElementRef<HTMLInputElement>;
+
+  maxLength = ISSUE_BODY_SIZE_LIMIT;
+
+  history: UndoRedo<textEntry>;
+
+  ngOnInit() {
+    if (this.initialTitle !== undefined) {
+      this.titleField.setValue(this.initialTitle);
+    }
+
+    if (this.titleField === undefined || this.titleForm === undefined || this.id === undefined) {
+      throw new Error("Title Editor's compulsory properties are not defined.");
+    }
+
+    this.titleField.setValidators([Validators.required, Validators.maxLength(this.maxLength)]);
+    this.history = new UndoRedo<textEntry>(
+      75,
+      () => {
+        return {
+          text: this.titleInput.nativeElement.value,
+          selectStart: this.titleInput.nativeElement.selectionStart,
+          selectEnd: this.titleInput.nativeElement.selectionEnd
+        };
+      },
+      500
+    );
+  }
+
+  onKeyPress(event: KeyboardEvent) {
+    if (this.isUndo(event)) {
+      event.preventDefault();
+      this.undo();
+      return;
+    } else if (this.isRedo(event)) {
+      this.redo();
+      event.preventDefault();
+      return;
+    }
+  }
+
+  onPaste(event: ClipboardEvent) {
+    // the text area is not changed at this point
+    this.history.forceSave(null, true, false);
+  }
```

Comment:
Originally in `comment-editor` there was additional code to handle paste events due to the possibility of users pasting in images. 

However, in the case of titles, I think we can combine this with either `beforeinput` or `input` event or both. 
This depends on how you want the behavior to be cached (up to you for this case since for pure text input the different won't actually be that noticeable)

Reply:
This is a good idea, the `onPaste` being separate in this case doesn't really do much. I think I'll combine this with the `beforeinput` event then, thanks for the suggestion!